devito/operator/operator.py

```diff
@@ -361,6 +364,9 @@ def _lower_clusters(cls, expressions, profiler=None, **kwargs):
         # Generate implicit Clusters from higher level abstractions
         clusters = generate_implicit(clusters, sregistry=sregistry)
 
+        # Lower all remaining high order symbolic objects
+        clusters = lower_index_derivatives(clusters, **kwargs)
```

Comment:
unnecessary verbosity, it seamlessly acts as a no-op if there are no IndexDerivatives

Reply:
It acts as it but it still needs to visit it and figure out if there is existing one, so would put an if either here or make expand a kwarg of `lower_index_derivatives` to immediately return without doing anything

Reply:
There are a few "buts" here...

1) The performance penalty of a clusters visit which ultimately acts as a no-op is completely neglible.

This is TTI so8 with the `lower_index_derivative` pass enabled (acting as a no-op obv).

```
Operator `ForwardTTI` generated in 8.58 s
``` 

Now the same with the entire pass commented out:

```
Operator `ForwardTTI` generated in 8.60 s
```

2 seconds more simply because it's in the noise. So virtually 0 overhead. 

2) In theory, one could bypass the abstraction and construct an Eq with an IndexDerivative, and this would result in a compilation error ultimately because it wouldn't get lowered

So, in general, I'm against the placement of "if"s to return immediately if something ain't supposed to happen. Think about MPI... all MPI passes are performed, including the data dependence analysis, even though ultimately, MPI doesn't get generated! The reason is that, for robustness, we should always attempt to perform all compiler passes to make sure their design is such that they can easily handle the fallback/no-op cases (amongst other obvious things)

With all this mind, I get that here I'm probably overkilling it -- you were just asking to add an extra `if` after all, and honestly speaking, if my arguments won't convince you, I'm OK with adding it in, just say it  :-)